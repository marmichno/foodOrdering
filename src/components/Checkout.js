import {useEffect, useState} from 'react';
import {quantityMinus} from '../actions/index'
import {quantityPlus} from '../actions/index'
import {useDispatch, useSelector} from 'react-redux';
import {Link} from 'react-router-dom';
import {motion} from 'framer-motion';
import {IoIosAddCircleOutline} from 'react-icons/io'

const containerVariants = {
    hidden: {
        x: -1000
    },
    visible: {
        x: 0,
        transition: {
            type: 'tween',
            duration: 0.5
        }
    },
    exit: {
        x:-1000,
        transition: {ease: 'easeInOut', duration:0.5}
    }
  }



export const Checkout = ({hideCheckout}) => {

    const [checkoutSum, setCheckoutSum] = useState(0);
    const dispatch = useDispatch();
    const order = useSelector(state => state.manageCheckoutReducer);

    useEffect(() => {
        if(order.length > 0){
            setCheckoutSum(order.map(value => value.price * value.quantity).reduce((a, b) => a + b))
        }
    },[order]);

    const changeQuantity = (e) => {
        const whichProduct = e.target.dataset.name;
        const removeOrAdd = e.target.innerHTML;

        if(removeOrAdd === '+'){
            dispatch(quantityPlus(whichProduct));
        }else{
            dispatch(quantityMinus(whichProduct));
        }
    }

    
    if(order.length > 0){
        return(
                <motion.div 
                variants={containerVariants}
                initial="hidden"
                animate="visible"
                exit="exit"
                className="orderToCheckout">
                    <div onClick={hideCheckout} className="hideCheckout"></div>
                    {order.map(value => {
                        return(
                            <div className="checkoutProduct">
                                <div className="productImage">
                                    <h1>{value.name}</h1>
                                </div>
                                <div className="productPriceData">
                                    <div>
                                        <p>Price:{(value.price * value.quantity).toFixed(2)}$</p>
                                    </div>
                                    <div>
                                        <button onClick={changeQuantity} className="plus" data-name={value.name}>+</button>
                                        <p>quantity:{value.quantity}</p>
                                        <button onClick={changeQuantity} className="minus" data-name={value.name}>-</button>
                                    </div>
                                </div>
                            </div>
                        )
                    })}
                    <div className="checkoutSum">
                        <div>
                            <hr></hr>
                        </div>
                        <h1>Subtotal: {checkoutSum.toFixed(2)}$</h1>
                        <Link to={{
                            pathname:'/order/choose_delivery'
                        }}>
                            <button>Go to order details</button>
                        </Link>
                    </div>
                </motion.div>
        )

    }else{
        return(
            <motion.div
            variants={containerVariants}
            initial="hidden"
            animate="visible"
            exit="exit"
            className="orderToCheckout">
                <IoIosAddCircleOutline className="x" onClick={hideCheckout}/>
                <div className="emptyContainer" onClick={hideCheckout}>
                    <h1>Your cart is empty :(</h1>
                    <hr></hr>
                    <p>add something to cart to continue</p>
                </div>
            </motion.div>
        )
    }
}