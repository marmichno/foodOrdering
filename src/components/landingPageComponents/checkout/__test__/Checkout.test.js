import Checkout from '../Checkout';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import * as redux from 'react-redux';
import {render, fireEvent, waitFor} from '@testing-library/react';
import {BrowserRouter} from 'react-router-dom';
import TestRenderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";


const store = createStore(() => [], {}, applyMiddleware());

test("checks message if store is empty", async () => {

    const spy = jest.spyOn(redux, 'useSelector');
    spy.mockReturnValue([])

    const {getAllByText} = render(<BrowserRouter><Provider store={store}><Checkout/></Provider></BrowserRouter>);

    await waitFor(() => {
        expect(getAllByText('Your cart is empty :('));
    })
});

test("test if product is rendered", async () => {

    const spy = jest.spyOn(redux, 'useSelector');
    spy.mockReturnValue([{ 
        price:"6.00",
        name:"philadelphia",
        quantity:1
     }])


    const {getByText} = render(<BrowserRouter><Provider store={store}><Checkout/></Provider></BrowserRouter>);

    await waitFor(() => {
        expect(getByText('philadelphia'));
    })
});


// test("check if + calls dispatch", async () => {
//     const spy = jest.spyOn(redux, 'useSelector');
//     spy.mockReturnValue([{ 
//         price:"6.00",
//         name:"philadelphia",
//         quantity:1
//      }])

//      const {getByTestId} = render(<BrowserRouter><Provider store={store}><Checkout/></Provider></BrowserRouter>);

//      store.dispatch = jest.fn();

//      await waitFor(() => {
//          fireEvent.click(getByTestId("plus"));
//      })
 
 
//      expect(store.dispatch).toHaveBeenCalledTimes(1);
// });

test("check if - calls dispatch", async () => {
    const spy = jest.spyOn(redux, 'useSelector');
    spy.mockReturnValue([{ 
        price:"6.00",
        name:"philadelphia",
        quantity:2
     }])

     const {getByTestId} = render(<Provider store={store}><BrowserRouter><Checkout/></BrowserRouter></Provider>);

     store.dispatch = jest.fn();

     await waitFor(() => {
         fireEvent.click(getByTestId("minus"));
     })
 
 
     await waitFor(() =>{
        expect(store.dispatch).toHaveBeenCalledTimes(1);
     })
});