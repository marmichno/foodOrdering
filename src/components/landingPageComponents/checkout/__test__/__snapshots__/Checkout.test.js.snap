// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="orderToCheckout"
  style={
    Object {
      "transform": "translateX(-1000px) translateZ(0)",
    }
  }
>
  <div
    className="hideCheckout"
  />
  <div
    className="checkoutProduct"
  >
    <div
      className="productImage"
    >
      <h1>
        philadelphia
      </h1>
    </div>
    <div
      className="productPriceData"
    >
      <div>
        <p>
          Price:
          12.00
          $
        </p>
      </div>
      <div>
        <button
          className="plus"
          data-name="philadelphia"
          data-testid="plus"
          onClick={[Function]}
        >
          +
        </button>
        <p>
          quantity:
          2
        </p>
        <button
          className="minus"
          data-name="philadelphia"
          data-testid="minus"
          onClick={[Function]}
        >
          -
        </button>
      </div>
    </div>
  </div>
  <div
    className="checkoutSum"
  >
    <div>
      <hr />
    </div>
    <h1
      data-testid="subtotalPrice"
    >
      Subtotal: 
      0.00
      $
    </h1>
    <a
      href="/order/choose_delivery"
      onClick={[Function]}
    >
      <button>
        Go to order details
      </button>
    </a>
  </div>
</div>
`;
